src/utils/daos_hdlr.c

```diff
@@ -899,6 +900,41 @@ fs_copy(struct cmd_args_s *ap, struct file_dfs *src_file_dfs, struct file_dfs *d
 		}
 	}
 
+	/* Make sure the source and destination are not the same.
+	 * Make sure the source is not a subset of the destination. */
+	if ((src_file_dfs->type == POSIX && dst_file_dfs->type == POSIX) ||
+	    (src_file_dfs->type == DAOS && dst_file_dfs->type == DAOS &&
+	     strcmp(ca->src_pool, ca->dst_pool) == 0 && strcmp(ca->src_cont, ca->dst_cont) == 0)) {
+		if (strcmp(src_path, dst_path) == 0) {
+			rc = -DER_INVAL;
+			DH_PERROR_DER(ap, rc, "source and destination are the same");
+			D_GOTO(out, rc);
+		}
+		if (S_ISDIR(src_stat.st_mode)) {
+			int  src_path_len;
+			bool src_is_subpath;
+
+			/* Always compare src_path with a trailing slash */
+			src_path_len = strlen(src_path);
+			if (src_path[src_path_len - 1] == '/') {
+				src_is_subpath = (strncmp(src_path, dst_path, src_path_len) == 0);
+			} else {
```

Comment:
What if src_path ends with '/' but dst_path does not end with '/'?
Example,
src_path = "/dir0/dir1/"
dst_path = "/dir0/dir1"
src_is_subpath is set false.

[leihuan1@boro-46 daos]$ cp -R /dev/shm/dir/ /dev/shm/dir
cp: cannot copy a directory, '/dev/shm/dir/', into itself, '/dev/shm/dir/dir'

Reply:
That behaves as expected
```
$ daos fs copy --src dir0/dir1/ --dst dir0/dir1
ERROR: cannot copy a directory into itself: Invalid parameters (-1003)
```
If the source doesn't end with `/`, we append it before comparing.
And because we always generate `dst_path` like this:
https://github.com/daos-stack/daos/blob/d4cc14fd97a7cc522669f7899622920202015ae1/src/utils/daos_hdlr.c#L894-L900
it will either contain a slash at the some location as the trailing slash on the source, or it will not be a subpath of the source

Reply:
I see. Thank you!